VCFv4.3.tex

```diff
@@ -503,41 +504,87 @@ \subsection{VCF tag naming conventions}
 
 \section{INFO keys used for structural variants}
 \label{sv-info-keys}
-The following INFO keys are reserved for encoding structural variants. In general, when these keys are used by imprecise variants, the values should be best estimates. When a key reflects a property of a single alt allele (e.g. SVLEN), then when there are multiple alt alleles there will be multiple values for the key corresponding to each alelle (e.g. SVLEN=-100,-110 for a deletion with two distinct alt alleles).
+The following INFO keys are reserved for encoding structural variants. When a key reflects a property of a single alt allele (e.g. SVLEN), then when there are multiple alt alleles there will be multiple values for the key corresponding to each alelle (e.g. SVLEN=-100,-110 for a deletion with two distinct alt alleles).
 \footnotesize
 \begin{verbatim}
 ##INFO=<ID=IMPRECISE,Number=0,Type=Flag,Description="Imprecise structural variation">
 ##INFO=<ID=NOVEL,Number=0,Type=Flag,Description="Indicates a novel structural variation">
 ##INFO=<ID=END,Number=1,Type=Integer,Description="End position of the variant described in this record">
 \end{verbatim}
 \normalsize
-For precise variants, END is POS + length of REF allele - 1, and the for imprecise variants the corresponding best estimate.
+For precise variants, END is POS + length of REF allele - 1.
```

Comment:
Please state what would be the preferred action in the case of imprecise variants, otherwise it will be open to interpretation.

Reply:
> The above comment would lead to a point that the `IMPRECISE` flag is useful

I am not clear why this is the case. To me it would seem to indicate the contrary: Since we have the exact haplotype and the variant is "precise", of what use is the `IMPRECISE` tag in this scenario? @Shuang-Broad, could you please elaborate?


> the deleted sequence is then any contiguous substring of the sequence `HOM + UNIQUE_DEL + HOM` with a size of `len(HOM) + len(UNIQUE_DEL)`. Sticking to the left-aligning convention would put the `POS` at the end of `LEFT_FLANK`.


Do I understand correctly that, in this case, `POS` would indicate the first nucleotide of `HOM`? Setting aside alternative arguments for right- or left-aligning for the moment, agnostically speaking it seems as though `POS` ought to correspond to the position at which the reference and test sample first diverge, i.e. the first base of `UNIQUE_DEL` (or the preceding base, per VCF convention). Would this be called right-aligning? In any case, I realize there may be other, stronger arguments that bear on the question of which is the correct solution.

Reply:
Sorry for the late reply, @thefferon .
1. You are right, the `IMPRECISE` tag is absolutely unnecessary for the example deletion case, because we have the exact haplotype, and I'm sorry for the confusion. What I wanted to say is that when the exact haplotype is not available (e.g. repetitive region, where only pair information is suggesting signal), the `IMPRECISE` tag is valuable.
2. Maybe a concrete example would be helpful in delivering what I mean: suppose the ref allele is `ATCG + AAA + GCGCGC + AAA + CCCCCCC`, and the alt is `ATCG + AAA + CCCCCCC`, then we can only infer that a contiguous substring with length 9 of `AAA + GCGCGC + AAA` is gone; left aligning the `POS` would fix it to the location of the base `G` in the left flanking sequencing `ATCG`, because the current standard requires the `POS` to be __a base before__ the deleted sequence.

Reply:
@SHuang-Broad ,

Thanks for your reply. I have a bit of follow-up:

1. Thanks for clarifying. My original point about the `IMPRECISE`tag (apart from the fact that it has never been defined in the spec) is: When there is ambiguity about the exact location of breakpoints of, e.g., a deletion, then CIPOS and CIEND should be used. And when they are used, they convey the message that the breakpoints are imprecise - and they are more useful than `IMPRECISE` because they provide at least some information about the extent of the ambiguity. Now, returning to your last example, do you still feel that `IMPRECISE` would be useful? Wouldn't you expect to see CIPOS and CIEND used in this case, and wouldn't that make `IMPRECISE` redundant?
2. I understand your point about left aligning: in your example it will never be possible to know whether the deleted sequence is `AAA GCGCGC`, `AA GCGCGC A`, `A GCGCGC AA`, or `GCGCGC AAA`. For sanity, there is clearly a need to use a convention in these cases; and I believe the two clearest options are 'left-align' and 'right-align'. Speaking in terms of logic, the decision as to which to use is arbitrary. But to use your example again I am proposing that, if you were to compare the reference seq to test seq, reading 5'-to-3', you wouldn't see a difference between them until you reached position '8', where the ref has 'G' and the test has 'C':
`ATCGAAAGCGCGCAAACCCCCCC`
`ATCGAAACCCCCCC`
 Thus in this sense it would seem logical to establish `POS` at position '7' (the third 'A') and to assert the deletion is `GCGCGC AAA`:
`ATCGAAAGCGCGCAAACCCCCCC`
`ATCGAAA---------CCCCCCC`
It is just now occurring to me to ask whether the term `IMPRECISE` might be needed here, to alert the reader to the fact there are multiple ways to interpret the call! But for the sake of normalization, not to mention consistency, a choice must be made between 'left-align and 'right-align'.

Reply:
1. Yes, I do feel that `IMPRECISE` is still useful, for imprecise variants, i.e. the variants that we don't have the exact haplotype, which isn't the case for the example deletion. I was trying to clarify the difference between (exact haplotype known, exact place with ambiguity) and (exact haplotype impossible with data, exact place impossible hence). The first case should have `CIPOS` and `CIEND` annotations, but not the `IMPRECISE` flag because the alt/sample haplotype is exactly known with ambiguity in how to align to the reference.
2. The example was, like intended, an example, so it didn't go through very stringent scrutiny, and might have edge case. But, the aligner might produce chimeric alignments for the sample haplotype when aligned to the reference sequence, with cigars `7M16S` when the alignment contains the left flanking sequence, and `11S12M` when the alignment contains the homopolymer `C`'s. In other words, an aligner could do this:
   ```
   ATCGAAAGCGCGCAAACCCCCCC
   ATCG---------AAACCCCCCC
   ```
3. (sorry for bringing up another item) I totally agree that for records in VCF, interpretations can sometimes be changed and leading to the same haplotype when one has too much time to kill (I remember seeing two VCF records, with essentially the same haplotype when reconstructed yet having very different interpretations&mdash;INS/DUP). I believe this point is echoed in my other comments: that the interpretations should always be viewed with a healthy suspicion and it's the haplotype (when available) that is the essence of the game. But this is not the same as the variant being "imprecise", because again, we do have the concrete haplotype a lot of times.

Man this discussion is getting longer and longer :wink:

Reply:
Take the following two variants:
```
ACGTAAAAAAACGT
ACGTA-----ACGT
     ^---^ (called position)
````
```
ACGTAACCTTGCGT
ACGT???-???CGT
     ^---^ (called position)
````

In the both cases, `CIPOS=-1,1`. In the first case, this is due to the inherit ambiguity in the breakpoint position due to microhomology. In the second case, the breakpoint is `IMPRECISE` due to lack of sequencing across the breakpoint (e.g. lacking split read support/called by a discordant read pair only caller)). In the latter, more sequencing could resolve the breakpoint exactly. VCF needs to be able to differentiate between these two scenarios and this is currently done using the  `IMPRECISE` flag.

Reply:
I think we are getting close to resolving this. But before I post at length:

@d-cameron , you wrote:
> In the latter, more sequencing could resolve the breakpoint exactly. In the former, more sequencing won't help as the breakpoint is already resolved exactly. VCF needs to be able to differentiate between these two scenarios and this is currently done using the `IMPRECISE` flag.

Is it your position, then, that the `IMPRECISE` flag should be applied in, and **_only_** in, the "latter" case - i.e. in which the variant lacks sequence across the breakpoint?

Another quick point about aligners and callers (in which you are both more versed than I).
@SHuang-Broad you wrote:
> ...the aligner might produce chimeric alignments for the sample haplotype when aligned to the reference sequence, with cigars `7M16S` when the alignment contains the left flanking sequence, and `11S12M` when the alignment contains the homopolymer C's. In other words, an aligner could do this:<pre><code>ATCGAAAGCGCGCAAACCCCCCC
ATCG---------AAACCCCCCC</code></pre>

And @d-cameron you gave the example:
<pre><code>ACGTAAAAAAACGT
ACGTA-----ACGT
     ^---^ (called position)
</pre></code>

which implies the caller chose this particular solution. Implicit in both these examples is the introduction, by the caller/aligner, of a **bias**; they have already selected one solution from among several equally valid solutions; and this affects downstream interpretation of the data. Each caller or aligner may introduce its own "interpretation". The VCF spec, and users of VCF data, would be much better off if such biases were eliminated at the level of the callers/aligners – by updating the software so it is aware when a problem has multiple solutions, each of which is equally plausible. Otherwise I see little hope of being able to work with the data in VCF systematically.

Reply:
In other words, any intelligent aligner, when presented with the following two sequences:
<pre><code>`ATCGAAAGCGCGCAAACCCCCCC`
`ATCGAAACCCCCCC'</pre></code>

should recognize – and encode in its output, in some form – that there are four equally valid interpretations that explain the difference between the sequences. Only then does it become possible to effectively make use of a left-align or a right-align rule.

Reply:
Furthermore, because all 4 solutions are equally valid, I don't think it makes sense to provide a `CIPOS` value without an 'align' rule - if you can't systematically identify a base to be `POS`, then you can't calculate `CIPOS`.

Reply:
> if you can't systematically identify a base to be POS, then you can't calculate CIPOS
As it currently stands, for exactly known SVs with breakend microhomology, implementations are free to choose any valid POS/CIPOS interval combination (GRIDSS chooses the middle position, other callers choose left, other use whatever the align did for the majority of the reads). There are already multiple ways to represent a SV in VCF so allowing multiple POS/CIPOS representation isn't a huge issue. It could be resolved by specifying a convention (e.g. left-aligning the lower breakend position (edge case: fold-back inversions)) but I'm not too fussed either way (my SV comparison implementation ignores the nominal POS and uses the interval (POS+CIPOS[0], POS+CIPOS[1]) for all comparison).

Reply:
>In other words, any intelligent aligner, when presented with the following ... should recognize – and encode in its output, in some form – that there are four equally valid interpretations that explain the difference between the sequences.

Alternatively, any intelligent variant caller would calculate any micro-homology intervals around the split read/indel positions reported by the aligner. GRIDSS, Socrates and Pindel already do this. Given that callers can't even trust aligners to write SAM files that actually make sense, my preference is to keep this in the VCF specifications and not spill over into the SAM as it's not absolutely necessary.

Reply:
>Is it your position, then, that the IMPRECISE flag should be applied in, and only in, the "latter" case - i.e. in which the variant lacks sequence across the breakpoint?

Correct.

The only edge cases I can think of are:

- when an RP caller has sufficient coverage that there is only one breakpoint position possible. One could argue that these calls are precise. I (weakly) prefer these to have `IMPRECISE` and `CIPOS=0,0` for consistency with your definition of precise calls requiring breakpoint sequence.

- when the full untemplated inserted sequence is not known. e.g: a ~300bp insertion with soft clip/breakpoint assembly support from both sides, spanning RP, but a fragmented breakpoint assembly. The full sequence is not known, but the breakpoint positions are exact. Should this be flagged as IMPRECISE? I'm inclined to say no since the breakend positions are in fact precise. The current VCF specs can actually represent such a call at all:
```
REF: ACGT------------------------------------CCGG
ALT: ACGTTTCAAA?????(50-100bp)????????TTTCCAACCGG
```
I guess you could put Ns in there, but using standard VCF fields, there's no way to say that there's somewhere between 50 and 100 Ns. Come to think of it, this is just a special case of an insertion of unknown length, something that the standard VCF fields don't support, merely a single SVLEN value.

Reply:
> Implicit in both these examples is the introduction, by the caller/aligner, of a bias; they have already selected one solution from among several equally valid solutions; and this affects downstream interpretation of the data. 

If the caller reports a POS and CIPOS, then it is reporting a set of solutions. If the downstream interpretation ignores most of these solutions then the issue is with the downstream interpretation, not what the caller reported. Just to reiterate: any tool that interprets VCF calls *MUST* be able to handle multiple different representations of the same variant. The only way that VCF readers can get away from this is if there is some sort of normalisation (similar to the left-alignment of SNV/indels). The only way I can see to normalise all SV events is to turn everything into left-aligned BND^, at which point, why have SVTYPE and any non-BND symbolic alleles at all?


^ This brings us back to the interpretation of SVTYPE. Is it a shorthand for the equivalent BND representation, or is it actually making a claim about the event type? Is a DEL caller claiming a CN reduction in the 'deleted' region? (for most callers, the answer is no). This is why I proposed the SVCLASS field. SVTYPE remains a restricted set of short-hand notations, and SVCLASS makes a claim about the mutation.

Reply:
As expressed in previous comments, when it comes to SV, what really matters, IMHO, are the haplotypes. Ideally, the tool that made the call set should provide the alt haplotype (ref is easy) explicitly when feasible (e.g. for complex SV) or provide a tool for doing this. So downstream analysis could use such information easily, for either querying the VCF to see if a variant is in close vicinity of a target region, or for making sense of what happened here. 

Now regarding the concerns for the "bias" of the aligner-given position, I don't think it is practical to require that from a general purpose aligner and special purpose aligner may provide no guarantee in these (homologous) regions either, because if we think about it, it is an optimization problem that also depends on the parameters sent to the aligner. But this should not stop us from doing the work: a good caller and downstream analysis in the SV world should be aware of this phenomenon and, echoing what has been said before, allow ambiguity and focus on the haplotype.

Reply:
I have a different opinion on the example insertion with unsure length that @d-cameron gave above.

It's getting clearer and clearer to me that when I refer to `IMPRECISE` variants, I am talking about "exact haplotype unknown", which is not the same as "exact place cannot be determined". In the "insertion with unknown length" example above, the exact alt sequence cannot be determined, hence the alignment could NOT be pinned down either (what if the missing sequence __can__ be aligned to the left or right flanking sequences?), making the alignment given in the example a bit suspicious.